### PR TITLE
TimeZoneInfo: Reduce intermediate allocations in GetTimeZoneIds on Unix

### PR DESCRIPTION
**Pull request by justinvp**
Avoid intermediate underlying array allocations as items are added to the resulting `List<string>` by specifying the capacity. Also, change the signature of the private method to return `List<string>` instead of `IEnumerable<string>` to avoid the enumerator allocation when enumerating the ids.

As an example, the zone.tab file on my OS X 10.11.6 machine has 445 lines.

It's easiest to review ignoring whitespace (I changed an if block to exit early, reducing the indentation): https://github.com/dotnet/coreclr/pull/8769/files?w=1

Note: For a future improvement, I'm thinking about replacing the use of `File.ReadAllLines` with a simple struct reader for reading the ids from the zone.tab file, which will allow deleting the private `StreamReader` dependency on Unix and reduce the number of string allocations (e.g. no need to create an array of strings for each line, only strings for each id).

### CODE REVIEW

File: src/mscorlib/src/System/TimeZoneInfo.Unix.cs [hunk 1]
```diff
@@ -240,38 +240,58 @@ private static IEnumerable<string> GetTimeZoneIds(string timeZoneDirectory)
...
+                return new List<string>();
+            }
+
+            // Estimate the number of ids in the file based on the total number of lines
+            // minus the number of comment lines at the beginning of the file (the tab file
+            // typically only has comment lines at the beginning of the file).
+            int estimatedTimeZoneIdCount = zoneTabFileLines.Length;
+            foreach (string zoneTabFileLine in zoneTabFileLines)
+            {
+                if (!string.IsNullOrEmpty(zoneTabFileLine) && zoneTabFileLine[0] == '#')
```
Comment (jkotas) [inline src/mscorlib/src/System/TimeZoneInfo.Unix.cs]: I am not sure whether this optimization to estimate the number of ids is worth the extra code. Using the number of lines as the initial list capacity should be almost as good, and I guess it is likely going to disappear anyway if you switch to more low-level parser.

**Comment by jkotas** [comment]
LGTM modulo one comment

File: src/mscorlib/src/System/TimeZoneInfo.Unix.cs (hunk 1 above)
Comment (justinvp) [inline src/mscorlib/src/System/TimeZoneInfo.Unix.cs]: Fixed

**Comment by jkotas** [comment]
Thanks!

**Pull request merged**